pkg/auth/providers/azure/conversions.go

```diff
@@ -0,0 +1,67 @@
+package azure
+
+import (
+	"errors"
+
+	"github.com/Azure/azure-sdk-for-go/services/graphrbac/1.6/graphrbac"
+	"github.com/manicminer/hamilton/msgraph"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func userToPrincipal(user interface{}) (*v3.Principal, error) {
+	var id, displayName, loginName string
+
+	switch u := user.(type) {
+	case *msgraph.User:
+		id = *u.ID
+		displayName = *u.DisplayName
+		loginName = *u.UserPrincipalName
```

Comment:
Are these ever nil?

Reply:
Since I handle errors in previous calls, when the user/group is initially fetched, I don't expect nils.

Reply:
I will add another nil check inside this function, just in case.

Reply:
Since we do exhaustive type checking here, if a user or group is nil, the default branch of the switch statement will be executed.